### ISSUE TITLE
Ingress left on AWS after helm_release resource destroyed

### ISSUE THREAD

**Issue by Opalo** [hashicorp/terraform-provider-helm#474]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version and Provider Version
Terraform v0.12.24
+ provider.aws v2.52.0
+ provider.helm v1.1.1
+ provider.kubernetes v1.11.1

### Provider Version
Helm
```
version.BuildInfo{Version:"v3.1.2", GitCommit:"d878d4d45863e42fd5cff6743294a11d28a9abce", GitTreeState:"clean", GoVersion:"go1.13.8"}
```
### Affected Resource(s)

- helm_release

### Terraform Configuration Files
```hcl
resource "helm_release" "graylog" {
  depends_on = [
    module.eks.ext_dns
  ]
  chart      = "stable/graylog"
  name       = "graylog"
  version    = "1.6.2"
  values     = [
    file("${path.module}/../../k8s/${local.environment}/graylog/values.yaml")
  ]
}
```

```yaml
graylog:
  rootPassword: lol
  env:
    GRAYLOG_HTTP_EXTERNAL_URI: graylog.lol.com
  image:
    repository: graylog/graylog:3.1.3
  ingress:
    enabled: true
    annotations:
      kubernetes.io/ingress.class: alb
      alb.ingress.kubernetes.io/scheme: internet-facing
      alb.ingress.kubernetes.io/listen-ports: '[{"HTTP": 80}, {"HTTPS": 443}]'
      alb.ingress.kubernetes.io/actions.ssl-redirect: '{"Type": "redirect", "RedirectConfig": { "Protocol": "HTTPS", "Port": "443", "StatusCode": "HTTP_301"}}'
    hosts:
      - graylog.lol.com
    extraPaths:
      - path: /*
        backend:
          serviceName: ssl-redirect
          servicePort: use-annotation
  replicas: 1
  service:
    type: NodePort
  input:
    tcp:
      service:
        annotations:
          service.beta.kubernetes.io/aws-load-balancer-internal: 0.0.0.0/0
          external-dns.alpha.kubernetes.io/hostname: input.graylog.lol.com
        type: LoadBalancer
      ports:
        - name: input
          port: 12201
```
### Debug Output

Unavailable

### Panic Output

None

### Expected Behavior

After calling `terraform destroy infrastructure`, graylog (installed with `helm_release`) should be removed and the whole AWS EKS cluster should be destroyed.

### Actual Behavior
From the output of terraform I can see that `helm_release.graylog` was indeed destroyed, however `terraform destroy infrastructure` failed with: 

```
Error: Error waiting for internet gateway (igw-whatever) to detach: timeout while waiting for state to become 'detached' (last state: 'detaching', timeout: 15m0s)
```

It happened, because after removing `helm_release.graylog` there was an ALB instance left.

It all work correctly with separate steps: `terraform apply (..)` with `helm install (..)` and `helm delete (..)` with `terraform destroy (..)`

### Steps to Reproduce

It seems that it's enough to use the `helm_release.graylog` with the configuration I provided.

### Important Factoids

Running on AWS EKS.

**Comment by zhernovskyi** [comment]
@Opalo If ALB instance was provisioned by some ingress controller(aws alb controller for instance), maybe that controller was deleted before graylog release destroy was started? Use dependency if so.

**Comment by Opalo** [comment]
@zhernovskyi I had the dependencies defined properly AFAIR. However I can't check it right now, since I've stopped using helm provider. It was extremely unstable in my case.

**Comment by advissor** [comment]
Hi, 
I got similar issue for EKS cluster and randomly found this topic
For me , everything provisioned inside kubernetes cluster was attempted to destroyed after cluster was removed
This left ingresses & sgs & helm_releases yelling 

So the fix was , prior to destroy run : **terraform refresh** 

Were I found the solution : https://github.com/terraform-aws-modules/terraform-aws-eks/issues/1162

Apparently in 0.15-beta1 this is already addressed

**Comment by dak1n1** [comment]
> @Opalo If ALB instance was provisioned by some ingress controller(aws alb controller for instance), maybe that controller was deleted before graylog release destroy was started? Use dependency if so.

I saw this comment and immediately thought of the AWS LoadBalancer Controller that I've used in the past. It did indeed have this behavior. I ended up removing it from one of our CI jobs for that reason. (I found it easier to use the [nginx ingress controller](https://github.com/hashicorp/terraform-provider-kubernetes/blob/795bb184bb0400b1198e3d5ed454fe0a42f904c2/kubernetes/test-infra/eks/kubernetes-config/main.tf#L59) instead, especially since it works across multiple clouds).

The OP's config implies that this controller may be in use: `kubernetes.io/ingress.class: alb`. If that is the case, then the controller is creating a load balancer (ALB) outside of Terraform. When Terraform goes to destroy the infrastructure, it doesn't know about the ALB that the controller created. When you delete an Ingress, such as during the `terraform destroy`, the controller uses a Finalizer to prevent the deletion from completing until it has removed the associated ALB. This works most times. However, when the controller's Deployment is deleted before this ALB deletion finishes, there's no mechanism left to remove the ALBs.

Maybe try using a `pre-delete` hook to ensure all your Ingresses have been deleted prior to removing the Deployment.
https://helm.sh/docs/topics/charts_hooks/

Unfortunately, in the helm provider, I don't think we have the ability to modify the order of deletion within a chart. That appears to be [handled in Helm](https://github.com/helm/helm/blob/6297c021cbda1483d8c08a8ec6f4a99e38be7302/pkg/releaseutil/kind_sorter.go#L71) itself. And while they do attempt to delete Ingresses before Deployments, it takes time for each ALB to be deleted, and my theory is that this process does not always fully complete before the Deployment is destroyed.
